lib/exercises.rb

```diff
@@ -1,29 +1,104 @@
+# frozen_string_literal: true
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
 
+# Time Complexity: O(n)
+# Space Complexity: O(n)
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+
+  new_hash = Hash.new()
+
+  strings.each do |word|
+    # calculate the hash key
+    key_word = word.downcase.chars.sort.join
+
+      if new_hash[key_word]  == nil
+        new_hash[key_word] = []
+      end
+      array_value = new_hash[key_word]
+
+    length = array_value.length
+    array_value[length] = word
+  end
+
+  return new_hash.values
 end
 
+    # Another solution
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+
+    # def grouped_anagrams(strings)
+    #   new_hash = Hash.new([])
+    #   strings.each.with_index { |word, i| new_hash[word.downcase.chars.sort.join] += [i] }
+    #   new_hash.map { |key, indexes| indexes.map { |i| strings[i] } }
+    # end
+
+
+
 # This method will return the k most common elements
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n^2)
+# Space Complexity: O(n)
+
 def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  return [] if list.empty?
+  hash1 = Hash.new()
+  # Create a hash where identical elements are grouped
+  # into array values and the element is a key
+  list.each do | num |
+    if !hash1[num]
+      hash1[num] = 1
+    else
+      hash1[num] += 1
+    end
+  end
+  # Put the elements in one array and the frequency of
+  # each element in another, sorted based on descending # freqency
+  elements = []
+  elements_frequency = []
+  hash1.each do | key, frequency |
```

Comment:
Could you sort the hash key-value pairs here or select the max k elements instead?